File: tests/scripts/reproduce_400.py

import logging
import asyncio
import sys
import os

# Fix import path
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

import httpx
from async_backend import AsyncZenAIBackend
from zena_mode.rag_pipeline import LocalRAG

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("Reproduction")


STRESS_TESTS = [
    # ("Normal Query", "What is ZenAI?"),
    # ("Large Context (2k tokens)", "Ignore previous." + " word" * 2000), 
    # ("Special Chars", "Hello \x00 World \uffff \U0001F600"), 
    ("Foreign Languages", "你好世界 (Hello World) | C'est la vie (That's life) | 🌮"), # UTF-8 Safety Check
]

async def reproduce():
    """Reproduce."""
    logger.info("🧪 Starting Stress Test...")
    backend = AsyncZenAIBackend()
    
    # Retry logic for health check... (omitted for brevity, assume present or safe to skip if we trust backend is up from prev run)
    
    async with backend:
        # 1. Stress Tests
        for name, prompt in STRESS_TESTS:
            logger.info(f"\n⚡ Testing: {name}")
            char_count = len(prompt)
            logger.info(f"📏 Size: {char_count} chars")
            
            try:
                # Manually construct payload to match async_backend's logic roughly, or use send_message_async
                # send_message_async constructs the full JSON.
                
                response_text = ""
                got_400 = False
                async for chunk in backend.send_message_async(prompt):
                    if "Server returned 400" in chunk:
                        logger.error(f"❌ CAUGHT 400 ERROR on {name}: {chunk}")
                        got_400 = True
                    else:
                        response_text += chunk
                
                if got_400:
                    logger.info("⚠️ Verified 400 Error Triggered!")
                elif response_text:
                    logger.info("✅ Success")
                else:
                    logger.warning("❓ No response, no error?")

            except Exception as e:
                logger.error(f"❌ Exception: {e}")


if __name__ == "__main__":
    asyncio.run(reproduce())

File: rust_output/src/reproduce_400.rs
use anyhow::{Result, Context};
use crate::async_backend::{AsyncZenAIBacken};
use crate::rag_pipeline::{LocalRAG};
use tokio;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static STRESS_TESTS: std::sync::LazyLock<Vec<serde_json::Value>> = std::sync::LazyLock::new(|| Vec::new());

/// Reproduce.
pub async fn reproduce() -> Result<()> {
    // Reproduce.
    logger.info("🧪 Starting Stress Test...".to_string());
    let mut backend = AsyncZenAIBackend();
    let _ctx = backend;
    {
        for (name, prompt) in STRESS_TESTS.iter() {
            logger.info(format!("\n⚡ Testing: {}", name));
            let mut char_count = prompt.len();
            logger.info(format!("📏 Size: {} chars", char_count));
            // try:
            {
                let mut response_text = "".to_string();
                let mut got_400 = false;
                // async for
                while let Some(chunk) = backend.send_message_async(prompt).next().await {
                    if chunk.contains(&"Server returned 400".to_string()) {
                        logger.error(format!("❌ CAUGHT 400 ERROR on {}: {}", name, chunk));
                        let mut got_400 = true;
                    } else {
                        response_text += chunk;
                    }
                }
                if got_400 {
                    logger.info("⚠️ Verified 400 Error Triggered!".to_string());
                } else if response_text {
                    logger.info("✅ Success".to_string());
                } else {
                    logger.warning("❓ No response, no error?".to_string());
                }
            }
            // except Exception as e:
        }
    }
}
